### PR TITLE
python3.pkgs.ml-dtypes: unpin build dependencies

### PR DESCRIPTION
**Pull request by tjni**
## Description of changes

Also disable some failing tests on aarch64-darwin.

This will help this package build once we switch from `pip` to [`build`](https://pypa-build.readthedocs.io/en/latest/). It looks like upstream chooses to pin these dependencies to their minor version, but it's a lot to keep it up to date in nixpkgs, so I chose to unpin it in a postPatch instead of fetching patches from upstream.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tjni** [comment]
@ofborg build python310Packages.ml-dtypes python311Packages.ml-dtypes

File: pkgs/development/python-modules/ml-dtypes/default.nix [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
-, setuptools
-, pybind11
+, absl-py
 , numpy
+, pybind11
 , pytestCheckHook
-, absl-py
+, setuptools
+, wheel
 }:
```
Comment (GaetanLepage) [inline pkgs/development/python-modules/ml-dtypes/default.nix:12]: Please order the inputs by order of appearance (except for `lib` to set first).

**Review by GaetanLepage — CHANGES_REQUESTED** [review]
Thank you for this fix !

File: pkgs/development/python-modules/ml-dtypes/default.nix [hunk 2]
```diff
@@ -27,9 +29,18 @@ buildPythonPackage rec {
     fetchSubmodules = true;
   };
 
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace 'numpy~=' 'numpy>=' \
+      --replace 'pybind11~=' 'pybind11>=' \
+      --replace 'setuptools~=' 'setuptools>='
+  '';
```
Comment (samuela) [inline pkgs/development/python-modules/ml-dtypes/default.nix:37]: why is this necessary/an improvement? in other words, if it ain't broke why fix it?

File: pkgs/development/python-modules/ml-dtypes/default.nix [hunk 3]
```diff
@@ -47,6 +58,14 @@ buildPythonPackage rec {
     rm -rf ./ml_dtypes
   '';
 
+  disabledTests = lib.optionals (stdenv.isDarwin && stdenv.isAarch64) [
+    # These tests have expected RuntimeWarning: invalid value encountered in cast
+    # that have not been ignored yet upstream
```
Comment (samuela) [inline pkgs/development/python-modules/ml-dtypes/default.nix:63]: what's the upstream issue tracking this? let's make sure to link to it

File: pkgs/development/python-modules/ml-dtypes/default.nix (hunk 2 above)
Comment (tjni) [reply pkgs/development/python-modules/ml-dtypes/default.nix:37]: This will break in `staging` now since https://github.com/NixOS/nixpkgs/pull/248866 was merged.

File: pkgs/development/python-modules/ml-dtypes/default.nix (hunk 3 above)
Comment (tjni) [reply pkgs/development/python-modules/ml-dtypes/default.nix:63]: I haven't found the time to research exactly if this is still an upstream issue or our issue. (https://github.com/jax-ml/ml_dtypes/pull/11 in v0.2.0 looks like it's supposed to fix these, but it doesn't.) I will try to find time to do this.

Comment (samuela) [reply pkgs/development/python-modules/ml-dtypes/default.nix:63]: yeah let's either fix on our end or report upstream and link to the upstream issue in a code comment
